x-pack/test/functional/apps/aiops/explain_log_rate_spikes_logs.ts

```diff
@@ -0,0 +1,303 @@
+      );
+
+      // Get the px values for the timestamp we want to move the brush to.
+      const { targetPx: targetBaselinePx } = await aiops.explainLogRateSpikesPage.getPxForTimestamp(
+        BASELINE_TS + DAY_MS / 2
+      );
+
+      // Adjust the right brush handle
+      await aiops.explainLogRateSpikesPage.adjustBrushHandler(
+        'aiopsBrushBaseline',
+        'handle--e',
+        targetBaselinePx + intervalPx * 10
+      );
+
+      // Adjust the left brush handle
+      await aiops.explainLogRateSpikesPage.adjustBrushHandler(
+        'aiopsBrushBaseline',
+        'handle--w',
+        targetBaselinePx - intervalPx * 10
+      );
+
+      // Get the new brush selection width for later comparison.
+      const brushSelectionWidthAfter = await aiops.explainLogRateSpikesPage.getBrushSelectionWidth(
+        'aiopsBrushDeviation'
+      );
+
+      // Assert the adjusted brush: The selection width should have changed and
+      // we test if the selection is smaller than two bucket intervals.
+      // Finally, the adjusted brush should trigger
+      // a warning on the "Rerun analysis" button.
+      expect(brushSelectionWidthBefore).not.to.be(brushSelectionWidthAfter);
+      expect(brushSelectionWidthAfter).not.to.be.greaterThan(intervalPx * 21);
+
+      await aiops.explainLogRateSpikesPage.assertRerunAnalysisButtonExists(true);
+
+      await ml.testExecution.logTestStep('rerun the analysis with adjusted settings');
+
+      await aiops.explainLogRateSpikesPage.clickRerunAnalysisButton(true);
+      await aiops.explainLogRateSpikesPage.assertProgressTitle('Progress: 100% — Done.');
+
+      // The group switch should be disabled by default
+      await aiops.explainLogRateSpikesPage.assertSpikeAnalysisGroupSwitchExists(false);
+
+      // Enabled grouping
+      await aiops.explainLogRateSpikesPage.clickSpikeAnalysisGroupSwitch(false);
+
+      await aiops.explainLogRateSpikesAnalysisGroupsTable.assertSpikeAnalysisTableExists();
+
+      const analysisGroupsTable =
+        await aiops.explainLogRateSpikesAnalysisGroupsTable.parseAnalysisTable();
+      expect(analysisGroupsTable).to.be.eql(testData.expected.analysisGroupsTable);
+
+      await ml.testExecution.logTestStep('expand table row');
+      await aiops.explainLogRateSpikesAnalysisGroupsTable.assertExpandRowButtonExists();
+      await aiops.explainLogRateSpikesAnalysisGroupsTable.expandRow();
+
+      const analysisTable = await aiops.explainLogRateSpikesAnalysisTable.parseAnalysisTable();
+      expect(analysisTable).to.be.eql(testData.expected.analysisTable);
+    });
+  }
+
+  describe('explain log rate spikes - artificial log data', function () {
+    this.tags(['aiops']);
+
+    before(async () => {
+      try {
+        await es.indices.delete({ index: ES_INDEX });
+      } catch (e) {
+        log.error(`Error deleting index '${ES_INDEX}' in before() callback`);
+      }
+      // Create index with mapping
+      await es.indices.create({
```

Comment:
I'd like to know more about why we need the data generation in the test code. An alternative could be to generate the data once and store the index as an esArchive, which could be loaded with a one-liner in all tests that need it.
If we decide to keep it as code, I'd suggest to move it to the `test_data` file, such that we have data generation and expected data closer together.

Reply:
Creating an esArchive would also be an option. This particular sample dataset would be ~1.2MB. Are you aware of any considerations regarding repository size when we add these?

My thoughts on creating it on the fly were: It's the first example doing this, but I was thinking esp. in the context of API integration tests that if this was a bit more generalized we could create these datasets on the fly for various use cases for Explain Log Rate Spikes without the need to maintain static datasets. Treating the datasets as code would allow us to more easily update them or create variations for additional tests without the need to create esArchives for each variation.

Some roughly measured stats while running the tests locally: Creating the index takes ~150ms, the code to create the docs for the dataset takes ~10ms, the bulk index request to then ingest all docs takes ~1 second. So the overhead to create the dataset on the fly is quite small. That is for the 8400 docs in the test. If I change that to create 84000 docs the dataset generation takes ~80ms and the bulk index request ~2 seconds. For comparison, restoring `farequote` (~86000 docs) via esArchive takes ~9s.

Reply:
Thanks for providing these insights.

> Are you aware of any considerations regarding repository size when we add these?

AFAIK there are no hard constraints, but the size is definitely something to keep in mind. In the past, we used to go with an upper limit of 4-5 MB, but smaller is better, so the ~1.2 MB wouldn't be an issue.

I see how the data generation from code gives us more flexibility for future testing. Happy to keep it. As mentioned above, I'd suggest to move it out of the test file, but that could also be done as part of the consolidation that you plan as part of the API integration tests.